### PR TITLE
Introduced deep children text nodes rerendering

### LINKED ISSUE #1093: Renderer does not re-render spaces in text nodes correctly
There are (at least) two cases when spaces are not correctly from view text nodes to dom text nodes after a change.

Those errors are caused by the fact, that during re-render only the changed dom text node is updated. https://github.com/ckeditor/ckeditor5-engine/pull/1083 changes the way spaces are converter to `&nbsp;`. Now, changing to `&nbsp;` bases on the data of the previous text node. This means that whenever a text node is changed, next touching text node has to be re-rendered, etc.

Mentioned cases:

Case 1:

1. DOM is `<p>x<strong> y</strong></p>`.
2. Type space after `x`.
3. DOM becomes `<p>x <strong> y</strong></p>`.
4. Only one space is rendered.

Case 2:

1. DOM is `<p><strong>&nbsp;y</strong></p>`.
2. Put caret before `&nbsp;` and turn off bold.
3. Type `x`.
4. DOM becomes `<p>x<strong>&nbsp;y</strong></p>`.
5. Incorrect `&nbsp;` stays in DOM.

We know that there is a "bug" in `DomConverter.viewChildrenToDom`. View elements mapped to DOM elements are not re-rendered. Those elements might have text nodes that need refreshing and this is the moment when the bug occurs. Unfortunately if we want to change `viewChildrenToDom` behavior, we will do a bit more processing. Hopefuly, this won't have negative impact on editor.

PS.
@Reinmar told me to hide those tests in case we need them: https://gist.github.com/scofalik/7d33adb45b82722bf5f6ca59ae8439d8

### LINKED ISSUE #1125: Rendering deeper changes fails to update text nodes
Example:

1. Starting view and DOM: `<p><a>Text<b>X</b></a></p>`.
2. Let's say that you type something at the end of this content which makes browser to create a lot of mutations and inserting the text in the wrong place (at the end of `<p>`) instead of at the end of `<b>` where we want to render it. It means that at this point we have the `<p>`'s children marked to be re-rendered.
3. Now, in the model everything is fine so we end up with such a view to be rendered: `<p><a>Text<b>XY</b></a></p>`.
4. What happens now is that the renderer tries to update first `<a>`'s children and then `<p>`'s children. It also wants to update "X" text node but since `<a>` will be rendered anyway, this is blocked.
5. `_updateChildren( <a> )` is executed but it takes `<b>` through `viewToDom()` which then uses `mapViewToDom()` (because in the view that's still the same element) and we get the old... `<b>X</b>` DOM element.
6. This leaves us with an "X" instead of "XY".

This needs to be fixed in `viewToDom()` – it seems that it needs to somehow update the entire tree (not only take the element from the mapper) or mark its contents to be re-rendered by the renderer after it does its job.

Causes the following issues:

* https://github.com/ckeditor/ckeditor5-engine/issues/1093
* https://github.com/ckeditor/ckeditor5-typing/issues/120

### LINKED ISSUE #1342: Rendering improvements
## Avoid unnecessary structure updates

Background: #1334.

How does selection conversion work?

1. MODEL: `<p><$text bold=true>xx</text></p>`  + `selection[bold=true]`
2. VIEW:
	1. `<p><b>xx</b></p>`
	2. `<p>[]<b>xx</b></p>` (after position conversion)
	3. `<p><B>[]</B><b>xx</b></p>` (during `writer.wrap( selection )` )
	4. `<p><B>[]xx</B></p>` (after attribute conversion; we're merging to the left, so a new element `<B>` is left)
3. RENDERING:
	* actual DOM: `<p><b>xx</b></p>`
	* expected DOM `<p><B>XX</B></p>` (we need to replace the entire `<b>` element; IDEA: don't replace `<b>` with `<B>` because they are similar!)

BUT... WHAT IF SOMETHING ELSE HAS CHANGED?

* actual DOM: `<p><b>xx</b></p>`
* expected DOM `<p><B>XXY</B></p>` (don't replace `<b>` with `<B>` but recursively go deeper to apply other possible changes, see #1125)

Possible solution: 

1. Make `diff( actual, expected, sameNode )` skip elements which have the same names and attributes (see `view.Element#isSimilar()`).
2. As a result, we won't replace `<b>` with `<B>`, which means that the view to DOM mapping will be outdated – the new view `<B>` element will be mapped to the new DOM `<B>` element which is not in the tree (because the old DOM `<b>` element was left there). Which means that we should update bindings after skipping such replacements. How to do that? Perhaps we can simply and blindly set new mappings after rendering: purge existing bindngs (in a limited scope), `updatedDomElements.forEach( mapToUpdatedViewElement )`.

--

## Deep structure updates

Background: https://github.com/ckeditor/ckeditor5-engine/issues/1125

1. `<p><a>Text<b>X[]</b></a></p>`
2. `<p><a>Text<b>X</b></a><b>Y[]</b></p>` (after typing "Y"; a lot mutations)
3. `<p><a>Text<b>XY</b></a></p>` (the model and the view after inserting "Y")

* ACTUAL DOM:		`<p><a>Text<b>X</b></a><b>Y[]</b></p>`
* EXPECTED DOM: 	`<p><a>Text<b>XY</b></a></p>`

---

## Fillers

1. Insert the inline filler when selection is at an inline element boundary. Makes sure that typed text will be inserted into the element we want it to be. Case: typing at the end of a link natively inserts text outside that link.
	* `<p>x<a>[]ZWS</a>x</p>`
	* `<p>x<a>yyyy[]ZWS</a>x</p>`
	* `<p>x<a>ZWS[]yyyy</a>x</p>`

    It's still unclear where should be selection – on the left or on the right side of the filler. Perhaps, it will depend on a case.
2. Do not touch fillers when composition takes place. Case: typing in empty `<b>`. Assuming that we don't do unnecessary DOM changes, removing the filler immediately after a letter was typed will always break the composition.

---

## Misc

* Use smarter text nodes methods to update them. Case: hard to tell. But feels like the right thing to do.
* Avoid selection rendering during composition. Case: may fix some issues where we incorrectly update model's selection during composition (may be frequent during more complicated autocorrect).

### PR DESCRIPTION
**Pull request by f1ames**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Other: Introduced deep children text nodes rerendering. Closes #1125. Closes #1093. Closes ckeditor/ckeditor5-typing#120. 

---

### Additional information

The proposed solution is very similar to what was initially proposed by @scofalik - https://github.com/ckeditor/ckeditor5-engine/issues/1125#issuecomment-332791847. The logic marking text nodes is placed in `renderer._updateChildren` as this is the place where element nodes are modified and based on that modifications we can decide which elements' text nodes should be considered for rerendered.

The mutation observer change is also described here - https://github.com/ckeditor/ckeditor5-engine/issues/1125#issuecomment-373376368.

### CODE REVIEW

**Comment by Reinmar** [comment]
@Mgsy Could you check this PR? There's a wide range of test cases so we should ba careful.

**Comment by f1ames** [comment]
This PR introduces mechanism which forces rerendering of sibling `text nodes` of the `text node` which was modified.

One visible issue of previous behaviour is described in #1093 - in general, renderer failed when changes in one `text node` required rerendering of some of its sibling (e.g. spaces rerendering in siblings). And the introduced change takes care of that. Here `renderer._updateText` function was not changed so rendering happens in the same manner (text is rerendered only if it different than existing ones). The only difference is that it is executed for more text nodes than before.

The other change which may have some impact is [change in rendering order](https://github.com/ckeditor/ckeditor5-engine/pull/1366/files#diff-4c87133ed830fc4ea0646426deba32cfL201). Previously, `text nodes` were rerendered first and `element nodes` after that. Now due to the logic which was applied (_rerender children `text nodes` of synced `element nodes`_), the order was changed so `element nodes` are rerendered before `text nodes`. So **it may affect any operation which does more complex changes (so updating text/elements or its attributes at the same time) and probably we should look for any regressions in such cases**.

The last thing was a change in [processing mutations](https://github.com/ckeditor/ckeditor5-engine/pull/1366/files#diff-57e194ad013563b5752afa147f447d67L209) already described in https://github.com/ckeditor/ckeditor5-engine/issues/1125#issuecomment-373376368. Here it affects how mutations are generated (previously `view element` was passed with its children to mutation event and now it's passed without it). **The scenarios which may get affected by this change are some bigger content modifications (when some bigger part of content, especially with deep structure, gets changed or partially updated) and also cases (I'm not sure if there are any of such situations) when editing one element inside editor changes the other one not in the subtree (so not a child but sibling or any not directly structurally connected element)**.

**Comment by Mgsy** [comment]
I've spent some time testing this one. It fixes mentioned issues and I haven't noticed any regressions caused by this fix.

**Comment by Reinmar** [comment]
There's one thing that would make us feel safer – a test which checks rendering one big batch of various type of changes – in text nodes, removing, adding elements in various places on various levels of a tree. The troublesome part is how text updates mix with element updates due to shifts in indexes.

**Comment by f1ames** [comment]
> a test which checks rendering one big batch of various type of changes – in text nodes, removing, adding elements in various places on various levels of a tree...

Makes perfect sense, I will try to create such test so we could check if the change in rendering order had some unexpected effects.

File: src/view/renderer.js [hunk 1]
```diff
@@ -531,6 +533,25 @@ export default class Renderer {
...
+	 */
+	_markDescendantTextToSync( viewElement ) {
+		if ( viewElement ) {
+			if ( viewElement.is( 'text' ) ) {
+				this.markedTexts.add( viewElement );
+			} else if ( viewElement.is( 'element' ) ) {
+				for ( const child of viewElement.getChildren() ) {
+					this._markDescendantTextToSync( child );
+				}
+			}
```
Comment (scofalik) [inline src/view/renderer.js]: I wonder when this else path is taken (what else could `viewElement` be), but CC is 100%.

Comment (f1ames) [reply src/view/renderer.js]: Theoretically it shouldn't be anything else so simple `if - else` should be sufficient. But checking it in both case is more future proof.

**Comment by f1ames** [comment]
@scofalik after merging `master` one unit test from model (looks irrelevant to the changes - `model/document/should call all already processed callbacks again if a callback returned true`) started to fail on Travis 🤔 But it seems `master` itself passes on Travis.

**Comment by scofalik** [comment]
For me it looks good. I've pushed some slight changes in docs.

**Comment by f1ames** [comment]
And I pushed tests @Reinmar requested, so now we just need to figure out why this one test is failing on Travis.

**Comment by f1ames** [comment]
This test also fails on `ckeditor5-engine#master` - https://travis-ci.org/ckeditor/ckeditor5-engine/builds/359328562#L2272 so it is not related to this PR.

**Pull request merged**